### PR TITLE
feat(metrics_indexer): Add rate limits functionality to indexer, take 2 [INGEST-1380]

### PR DESCRIPTION
**Pull request by untitaker**
This reverts commit a65c78a.

Reverts https://github.com/getsentry/sentry/pull/36524
Reapplies https://github.com/getsentry/sentry/pull/36263 + typing fixes

## Split up in the following PRs

Merge in this order (it should be safe to deploy after each merge)

- [x] https://github.com/getsentry/sentry/pull/37168
- [x] https://github.com/getsentry/sentry/pull/37170
- [x] https://github.com/getsentry/sentry/pull/37171
- [x] https://github.com/getsentry/sentry/pull/37173
- [ ] https://github.com/getsentry/sentry/pull/37174

### CODE REVIEW

**Review by fpacifici — COMMENTED** [review]
Since this is the second time we deplyo a change to the processing phase of the indexer and we miss a bug in testing, I believe we have two problems:
- the abstraction in the indexer is not right so it makes it hard to make changes with confidence.
- we lack unit tests.

I have an idea to solve both, and I think it would be worth to write another PR before this one to implement the idea. It should not be too big:
Basically what happens [here](https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/consumers/indexer/processing.py#L63-L239) is:
- for each batch
- extract all the strings to resolve from the batch
- talk to the indexer to resolve those strings
- validate data is ok
- fit the integers into the batch 
- compose the kafka messages.

What if we had a smarter `Batch` abstraction that takes care of the transformation:
```
class MessageBatch: # now this is jsut a List
     def __init__(List[Messages]):
         ....

     def extract_strinags(self) -> SomeStringsBag like `KeyCollection` 

     def map(self, KeyResult) -> MessageBatch[str]
          """
          This produces a new batch (actually modifying the strings in place) 
          with the mapped strings
          """
```

So the processor coordinates the work. It asks the batch to provide the strings. It validates them, then taks to the indexer to resolve them and record. Then it talks again to the batch to produce a new output batch with the results. The batch data structure knows the message payload, it can extract strings and map them back.

Now the batch data structure is stateless and can be unit tested to death.
What do you think ?

**Comment by untitaker** [comment]
@fpacifici We have integration tests that ended up not testing anything because they are using the mock indexer (which doesn't have rate limits) together with mocked/stubbed fetch results coming from the rate limiter, and those went out of sync with the actual implementation once I added FetchTypeExt in the final round of reviews.

We want to be able to test message processing and rate limiting together, i.e. their interaction, because that's the boundary where things broke. At the same time, if we want to test this integration to death we ideally find a way to do it without hitting the database in order to keep the testsuite fast. So this suggests to me that we should figure out a way to detach rate limiting (and maybe caching as well) from the postgres_v2 backend, so it can be tested together with message processing without hitting the DB.

So let's say we build this Batch abstraction, and if I understand you right it would not handle rate limiting, and the responsibilities of `indexer.bulk_record` would stay the same. Then we can unittest that to death, but as usual those unittests will need to simulate whatever can be returned from the indexer (and that can go out of sync with the actual implementation again), or they hit the real postgres indexer (in which case the tests will be very slow). That's fundamentally the same problem we're dealing with already.

lmk what you think

**Comment by fpacifici** [comment]
I think the two approaches are not incompatible. 
You are right we need a better way to test what comes from the indexer and rate limiter.
Still testing the processing logic in isolation is not supposed to be alternative to that. 
While this time the issue was related with the data returned by the indexer, last time it was pure processing. 

But I would be fine if you do not want to break it apart within this task.

**Comment by fpacifici** [comment]
Another idea to make it harder to break and easier to test 

Why don't we return a properly structured type here, the current state feels like a case of Primitive Obsession  :
https://github.com/getsentry/sentry/pull/36531/files#diff-4d832960b67c8c52695fce355ad53704c4c0bab03ef0fdd5c487817d383094edR162


If that returned a `Mapping[str, Metadata]` where `Metadata` is a dataclass or a named tuple with properties instead of a bare tuple, the type checker should have noticed the issue. We would not be able to do tuple unpacking. 
That `KeyResult` originally returned `Mapping[str, int]` so fairly harmless. Then we added metadata, then FetchType and then FetchTypeExt. Now it is likely abusing that original Mapping.

**Comment by untitaker** [comment]
@fpacifici had to do a double-take on your first post, if I understood it right then it's actually really simple and https://github.com/getsentry/sentry/pull/36669 is basically it

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 1]
```diff
@@ -1,29 +1,16 @@
...
+from sentry.sentry_metrics import indexer
 from sentry.sentry_metrics.configuration import UseCaseKey
 from sentry.sentry_metrics.consumers.indexer.batch import IndexerBatch
 from sentry.sentry_metrics.consumers.indexer.common import MessageBatch
+from sentry.utils import metrics
 
 logger = logging.getLogger(__name__)
 
 
-@functools.lru_cache(maxsize=10)
```
Comment (untitaker) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:13]: I got rid of those because otherwise the return type of `indexer` is `Any` in `process_messages`, and that would infect return values of `indexer.bulk_record`, and `get_fetch_metadata` (where i shipped a bug/crash last time)

File: tests/sentry/sentry_metrics/test_multiprocess_steps.py [hunk 2]
```diff
@@ -260,8 +260,7 @@ def __translated_payload(
     return payload
 
 
-@patch("sentry.sentry_metrics.consumers.indexer.processing.get_indexer", return_value=MockIndexer())
```
Comment (untitaker) [inline tests/sentry/sentry_metrics/test_multiprocess_steps.py:263]: since https://github.com/getsentry/sentry/pull/36397 this is no longer necessary. the global indexer is mocked and cleared on every testrun because of the `sentry_metrics` marker, because this exact decorator was scattered across hundreds of product/api tests as well it seemed like a good default

File: src/sentry/runner/commands/run.py [hunk 3]
```diff
@@ -552,7 +552,7 @@ def ingest_consumer(consumer_types, all_consumer_types, **options):
 @click.option("--input-block-size", type=int, default=DEFAULT_BLOCK_SIZE)
 @click.option("--output-block-size", type=int, default=DEFAULT_BLOCK_SIZE)
 @click.option("--factory-name", default="default")
-@click.option("--ingest-profile")
+@click.option("--ingest-profile", required=True)
```
Comment (fpacifici) [inline src/sentry/runner/commands/run.py:555]: Could you please double check that we are passing the ingest profile in prod for Release health before merging ?

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 4]
```diff
@@ -138,8 +138,8 @@ def extract_strings(self) -> Tuple[Mapping[int, Set[str]], Set[str]]:
     @metrics.wraps("process_messages.reconstruct_messages")
     def reconstruct_messages(
         self,
-        mapping: Mapping[int, Mapping[str, int]],
-        bulk_record_meta: Mapping[str, Tuple[int, FetchType]],
+        mapping: Mapping[int, Mapping[str, Optional[int]]],
+        bulk_record_meta: Mapping[str, Metadata],
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:142]: Not sure I get why this is not broken by org. 
The same string can be in two orgs and we could be able to write it in one and not in the other.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 5]
```diff
@@ -161,26 +161,81 @@ def reconstruct_messages(
...
+                        else:
+                            exceeded_org_quotas += 1
+                        continue
+
+                    new_tags[str(new_k)] = new_v
             except KeyError:
                 logger.error("process_messages.key_error", extra={"tags": tags}, exc_info=True)
                 continue
 
+            if exceeded_org_quotas or exceeded_global_quotas:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:201]: Please write a datadog metric here as well. The error is likely going to be dropped (we do not need sentry errors when a customer floods us) though the metric will still be useful

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 6]
```diff
@@ -161,26 +161,81 @@ def reconstruct_messages(
...
+                    fetch_types_encountered.add(metadata.fetch_type)
+                    output_message_meta[metadata.fetch_type.value][str(metadata.id)] = tag
 
             mapping_header_content = bytes(
                 "".join([t.value for t in fetch_types_encountered]), "utf-8"
             )
             new_payload_value["tags"] = new_tags
-            new_payload_value["metric_id"] = mapping[org_id][metric_name]
+            new_payload_value["metric_id"] = numeric_metric_id = mapping[org_id][metric_name]
+            if numeric_metric_id is None:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:225]: Same as above, let's have a metric

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:58]: Why do we need to collect metrics out of this method and the settings one? They are supposed to have a minimal impact, as they only build data structures from the input provided and have no external dependencies.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:80]: Please make this a NamedTuple or it will be impossible to manage.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisSlidingWindowRateLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
+
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[Tuple[KeyResult, FetchType, FetchTypeExt]]]:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:97]: A tuple containing a Tuple and Sequence of tuples should have a proper structured type. Otherwise it will be impossible to maintain.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        2. A key collection containing all unmapped keys that passed through the
+          rate limiter.
+
+        3. All unmapped keys that did not pass through the rate limiter.
+        """
+
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:115]: Why do you need to make a copy of the mapping ?

Also why is this called `granted_key_collection` ? This is the original mapping. It includes both allowed and rejected keys.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:61]: Why not returning a mapping? It seems it would be a lot more readable.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                                    for quota in grant.reached_quotas
+                                ),
+                            ),
+                        )
+                    )
+
+            granted_key_collection[org_id] = allowed_strings
+
+        state = use_case_id, requests, grants, timestamp
+        return state, KeyCollection(granted_key_collection), dropped_key_results
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:140]: Why do we need to return the use case id and the requests ?

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                )
+            )
+
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
+
+
+class WritesLimiter:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:83]: Could you please add some unit tests for this class? It is complex enough that we can mess up corner cases.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 14]
```diff
@@ -138,8 +138,8 @@ def extract_strings(self) -> Tuple[Mapping[int, Set[str]], Set[str]]:
     @metrics.wraps("process_messages.reconstruct_messages")
     def reconstruct_messages(
         self,
-        mapping: Mapping[int, Mapping[str, int]],
-        bulk_record_meta: Mapping[str, Tuple[int, FetchType]],
+        mapping: Mapping[int, Mapping[str, Optional[int]]],
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:141]: Please let's have a unit test for this method. 
There are at least 5 code paths to cover:
- all good, all strings resolved
- metric id rate limited 
- some tag keys rate limited
- some tag values rate limited
- the same string is successful in an organization and not in another (two different messages). 

While the end to end test for the `process` method is good and useful, you would not test each and every one of those code paths above in an end to end test. This is one of the advantage of breaking this logic out of the massive process method we had before. Now we can test each case separately without going crazy.

**Review by fpacifici — COMMENTED** [review]
Please see my comments in line.

Also I would consider splitting this PR. There is a lot going on. 
What about introducing the rate limiter without taking the results into account first ? Which means make the changes to hte sliding window rate limtier, add the `ratelimiters.py` class and wire it into the indexer, but still write all the strings on the DB and do not change anything in the reconstruct ? 
This way we would reduce the risk by introducing the rate limiter logic without impacting the messages in any way.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py:142]: it's a long-standing bug. I pointed it out in sns channel a few months ago. we should fix it, thought somebody else was going to do it. will do now

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 7 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:58]: yes ideally they would, but as we saw with reconstruct-messages, python code can have some significant overhead

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 11 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:61]: this was already a topic last time I attempted to merge this patchset into master, I wanted to avoid unnecessary type conversions between dict and list.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 8 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:80]: made it dataclass, done

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 10 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:121]: it doesn't. in the last line of the for-loop we write back the list of strings that fit into the quota.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 12 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py]: this `state` is an "opaque" value that the indexer shouldn't poke at and pass to `apply_write_limits`. I thought this would be the best way to ensure we read and write with the same usecase.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 9 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py]: I moved the innermost tuple to a dataclass, lmk if that is sufficient

**Comment by untitaker** [comment]
@fpacifici thanks for the thorough review and for keeping coding standards high. I think I have addressed all review comments.

I can split up the PR. Should I do this before or after you review the recent changes?

**Comment by fpacifici** [comment]
> I can split up the PR. Should I do this before or after you review the recent changes?

Sorry I did not see this comment before. I think it is good to split. No need for another round. Though I will try to do another review tonight before you will wake up. 
If you do not see it by the time you start working, please go ahead.

**Comment by untitaker** [comment]
I split up the PR into five different ones and edited the PR description to reflect in which order they need to be merged. After that I'll rebase this PR and hopefully the diff empty, if not I guess I'll just merge it.
